### ISSUE TITLE
Babel 8 Release Plan

### ISSUE THREAD

**Issue by nicolo-ribaudo** [babel/babel#10746]
We plan to release a new major version in the beginning of 2020.

This release won't have all the migration pain which there was while migrating from Babel 5 to Babel 6 and then from Babel 6 to Babel 7. We plan to only make a few breaking changes, and provide an easy migration strategy for each of them. Because of this, you won't see ~80 prereleases like we did in for Babel 7.0.0, but we plan to only release a few of them.

Babel 8.0.0 will _only_ contain breaking changes: we will release a minor version the same day, containing all the bug fixes and new features that would otherwise be released in 8.0.0.

This document is still a work in progress, but you can already start by applying the suggested migration strategies to your own codebase.

- ### Compilation breaking changes
  - [Don't remove uninitialized class fields when using Flow/TS](https://github.com/babel/babel/issues/10039) (pr for flow: [#10120](https://github.com/babel/babel/pull/10120))
    - **Area**: Flow and TypeScript transforms
    - **Impact**: High (only for flow and TypeScript users)
    - **Migration**:
      Flow users can already update their code to use the flow comments syntax if they don't want to initialized to `undefined`. This is officially [recommended](https://github.com/facebook/flow/issues/6811#issuecomment-496808192) by the Flow team:
      ```js
      class A {
        foo: ?string; // initialized to undefined
        /*:: bar: number */ // type-only
      }
      ```
      
      TypeScript users can use the new `declare` syntax, which introduced in TypeScript 3.7 and Babel 7.7:
      ```js
      class A {
        foo: string | void; // initialized to undefined
        declare bar: number; // type-only
      }
      ```
      Note that while this syntax is enabled by default in `@babel/parser` when using the `typescript` plugin, if you are using `@babel/plugin-transform-typescript` or `@babel/preset-typescript` you must enable the `declareFields` option:
      
      ```json5
      {
        presets: [
          ["@babel/typescript", { "declareFields": true }]
        ]
      }
      ```
  - [Require `@babel/plugin-proposal-dynamic-import` when transforming `import()` to SystemJS](https://github.com/babel/babel/blob/78cd63d9cfcd96e6a151c58fed392c3ee757d861/packages/babel-plugin-transform-modules-systemjs/src/index.js#L183-L185)
    - **Area**: `@babel/plugin-transform-modules-systemjs`
    - **Impact**: Medium
    - **Migration**: Add `@babel/plugin-proposal-dynamic-import` to your config: you can already do it in Babel 7. If you are using `@babel/preset-env`, you don't need to do anything.
    - **Notes**: All the other plugins which support dynamic import (`transform-modules-commonjs` and `transform-modules-amd`) require the separate plugin since it was introduced. We couldn't change it for `transform-modules-systemjs` because that package did already support dynamic import.
  - [Disallow sequence expressions inside JSX attributes](https://github.com/babel/babel/issues/8604) (pr: [#8787](https://github.com/babel/babel/pull/8787))
    - **Area**: JSX, `@babel/parser`
    - **Impact**: Low
    - **Migration**: If you are using them, you can already wrap them in parentheses:
      ```js
      <div key={foo, bar}></div> // Invalid
      <div key={(foo, bar)}></div> // Valid
      ```
  - [Transforms JSX spread properties using object spread](https://github.com/babel/babel/issues/9652)
    - **Area**: JSX
    - **Impact**: Medium
    - **Migration**: You can already have this behavior by using the `useSpread` option in Babel 7.7.0. If your code needs to run in an environment which doesn't support object spread, you can either use `@babel/preset-env` (recommended) or `@babel/plugin-proposal-object-rest-spread`. If you want to transpile `Object.assign` down to Babel's `_extends` helper (which is the current default behavior) you also need to enable `@babel/plugin-transform-object-assign`.

- ### API breaking changes
  - [Drop support for node <10.9](https://github.com/babel/babel/issues/9981)
    - **Area**: Every package
    - **Impact**: Medium
    - **Notes**: The 8.x Maintenance LTS cycle is currently scheduled to expire early on December 31, 2019 to align with the scheduled End-of-Life of OpenSSL-1.0.2. 10.9 is the first Node.js 10.x version with V8 6.8.
    - **TODO**: Upgrade `chokidar`, `find-cache-dir`, `globals`, `make-dir`, `slash`
  - [Don't generate `TSParenthesizedType` unless `createParenthesizedExpression` is enabled](https://github.com/babel/babel/issues/9546)
    - **Area**: `@babel/parser`
    - **Impact**: Low
    - **Migration**: If you need informations about parentheses, you can already enable the `createParenthesizedExpression` parser option (supported since Babel 7.4.0)
  - [Remove `t.jSX*` and `t.tS*` builder aliases](https://github.com/babel/babel/issues/6989) (pr: [#6993](https://github.com/babel/babel/pull/6993))
    - **Area**: `@babel/types`
    - **Impact**: Low
    - **Migration**: Use `t.jsx*` and `t.ts*` instead, which have been supported since Babel 7.0.0
  - [Reject invalid identifiers in `t.identifier` builder](https://github.com/babel/babel/pull/10621) (pr was reverted)
    - **Area**: `@babel/types`
    - **Impact**: Medium
    - **Notes**: Breaks `metro` bundler as of Nov 2019 ([#10645](https://github.com/babel/babel/issues/10645)) 
  - [Remove `jsonCompatibleStrings` generator option](https://github.com/babel/babel/issues/9943) (pr: [#9958](https://github.com/babel/babel/pull/9958))
    - **Area**: `@babel/generator`
    - **Impact**: Medium
    - **Migration**: `@babel/generator` allows to specify options for [jsesc](https://github.com/mathiasbynens/jsesc), the library used to escape printed values.
      If you are using the `jsonCompatibleStrings` option, you can replace it with `jsescOption: { json: true }`.
  - [Disallow `useBuiltIns: 0` when transforming JSX](https://github.com/babel/babel/pull/10572#discussion_r336564779)
    - **Area**: `@babel/plugin-transform-react-jsx`, `@babel/preset-env`
    - **Impact**: Low
    - **Migration**: Use `useBuiltIns: false` instead.

- ### Misc breaking changes
  - Bump peer dependency on `@babel/core` to `^8.0.0`
    - **Area**: Every package
    - **Impact**: None if you update every `@babel/*` package
  - [ESM runtime helper files should have the .mjs extension
](https://github.com/babel/babel/issues/8462)
    - **Area**: `@babel/runtime`, `@babel/runtime-corejs2`, `@babel/runtime-corejs3`
    - **Impact**: Medium
    - **Notes**: ES Modules are now unflagged ([nodejs/node#29866](https://github.com/nodejs/node/pull/29866)), and `.js` modules don't work with native `import` unless our `package.json` specifies `type: "module"` (which will break cjs helpers).

- ### Other _possibly_ breaking changes
  - [Remove ts type imports on Program:exit](https://github.com/babel/babel/issues/9991) (pr: [#10009](https://github.com/babel/babel/pull/10009))
    - **Area**: `@babel/plugin-transform-typescript`
    - **Impact**: Low
  - [Allow skipped `NodePath`s to be requeued](https://github.com/babel/babel/blob/43b623c1f1e86e6fb86cae8d955a84fd924380a4/packages/babel-traverse/src/path/context.js#L241-L247)
    - **Area**: `@babel/traverse`
    - **Impact**: Low
    - **Notes**: This is actually a bugfix, but it causes an infinite loop in the tdz implementation of `@babel/plugin-transform-block-scoping`
  - Look for comments containing "Babel 8" in the source code
    - **Area**: Every package
    - **Impact**: Low
    - **Notes**: Most of those comments are just for internal dependencies between packages. Any significant change will have a dedicated point in this list of breaking changes.

**Comment by TrejGun** [comment]
> Remove ts type imports on Program:exit 

will this allow generate .d.ts files?

**Comment by nicolo-ribaudo** [comment]
@TrejGun It is totally unrelated I think. I'd be happy to hear more about it on our Slack!

**Comment by fwh1990** [comment]
Wii 8.x be faster than 7.x when building my own project?

**Comment by nicolo-ribaudo** [comment]
No. Version 8.0 will only contain breaking changes, so if there are any performance improvements that we can make they will also be included in 7.8.0.

**Comment by Andarist** [comment]
I would like to propose a breaking change. 

I believe that `@babel/runtime` helpers should not get special treatment when transpiled to CJS - at least not for their export shape (used requires etc can still be optimized). I propose such output:
```diff
+"use strict";
+exports.__esModule = true;
+exports.default = _toPropertyKey;

-var _typeof = require("../helpers/typeof");
+var _typeof = require("../helpers/typeof").default;

-var toPrimitive = require("./toPrimitive");
+var toPrimitive = require("./toPrimitive").default;

function _toPropertyKey(arg) {
  var key = toPrimitive(arg, "string");
  return _typeof(key) === "symbol" ? key : String(key);
}

-module.exports = _toPropertyKey;
```

With such ESM-compatible output we could try to explore removing `useESModules` option from `@babel/plugin-transform-runtime` because it's possible to create a "proxy" `package.json`  that would allow bundlers & node to pick up the correct version of a particular helper on their own using a **single** path (that would be inserted by transform runtime). The directory structure could look smth like this
```js
helpers/toPropertyKey/package.json
helpers/toPropertyKey/toPropertyKey.js
helpers/toPropertyKey/toPropertyKey.mjs
```
with `package.json` being:
```json
{
  "name": "@babel/runtime/helpers/toPropertyKey",
  "private": true,
  "type": "module",
  "main": "./toPropertyKey.cjs.js",
  "module": "./toPropertyKey.js",
  "exports": {
    ".": {
      "require": "./toPropertyKey.cjs.js",
      "default": "./toPropertyKey.js"
    }
  }
}
```

**Comment by XaveScor** [comment]
@babel/preset-env using core-js@2 by default. Can @bable/preset-env@8 using core-js@3 only?

**Comment by nicolo-ribaudo** [comment]
Yeah, this is a good idea!
@zloirock WDYT about dropping support for `core-js@2`?

**Comment by zloirock** [comment]
Yes, `core-js@2` support should be dropped and the version of `core-js` should be required argument with `useBuiltIns` in `preset-env`. Maybe it could be good also to enforce specifying of minor `core-js` version.

**Comment by zloirock** [comment]
Why it should be dropped - copy-paste from the draft of one post:

> `core-js@3` was published 10 months ago - I hoped that this is enough to update. However, `core-js@2` still is used more often than `core-js@3` - if some dependencies depend on `core-js@2`, users don't wanna have some copies of `core-js` in the bundle. However, we should kill `core-js@2` ASAP. The main reason is even not some bugs, it's... V8 (de)optimizations - even if nothing is polyfilled. WTF? Now, if V8 saw the usage of some non-often used features ([like `@@species` pattern](https://github.com/zloirock/core-js/issues/677)), for example, in the features detection - V8 always will use the non-optimized version of a method that theoretically could use it. In `core-js@2`, that affects regexes, iterators, some array methods, typed arrays, promises, etc... Sometimes that causes ~100x performance degradation, let's say thanks to V8. At this moment, V8 is the most popular JS engine, so it's a complex of critical issues. Workarounds for a big part of those deoptimizations were added in `core-js@3.0`, for remaining - [at the latest patch releases](https://github.com/zloirock/core-js/issues/679). I don't see any reason to spent many days fixing those (and other) issues in the obsolete version.

**Comment by Pranav2612000** [comment]
Hey. I would love to contribute to Babel 8. What are some easy features I can work on? I think I can work on :- 
 

> Look for comments containing "Babel 8" in the source code
> Area: Every package
> Impact: Low
> Notes: Most of those comments are just for internal dependencies between packages. Any significant change will have a dedicated point in this list of breaking changes.

Are there any other issues I can contribute to and help?

**Comment by nicolo-ribaudo** [comment]
Another good one could be https://github.com/babel/babel/issues/9546, which would give you the chance to dig into `@babel/parser`'s internals.
Or maybe `Disallow using babel.transform, babel.transformFile, babel.transformFromAst, babel.parse, babel.loadOptions and babel.loadPartialConfig synchronously`, which is probably a bit easier.

If you want to work on the `TODOs left in the comments containing "Babel 8"`, please don't resolve all of them in a single PR but keep separate changes in different PRs.

**Comment by Pranav2612000** [comment]
> Disallow using babel.transform, babel.transformFile, babel.transformFromAst, babel.parse, babel.loadOptions and babel.loadPartialConfig synchronously
> Area: @babel/core
> Impact: Medium
> Migration: You can use babel.transformSync, babel.transformFromAstSync and babel.transformFileSync, supported since version 7.0.0. babel.parseSync, babel.loadOptionsSync and babel.loadPartialConfigSync will be introduced in v7.8.0.

Shoud I just delete the given functions? Or am I misunderstanding the requirements?

**Comment by nicolo-ribaudo** [comment]
Currently they behave asynchronously if you pass them a callback, otherwise they behave synchronously. Only the callback-based version should be allowed.

**Comment by nicolo-ribaudo** [comment]
For whoever is watching this issue:
I just added the `Use the new JSX implementation by default` section.

**Comment by nicolo-ribaudo** [comment]
Update: I added the `Parse JSX elements when both the JSX and TypeScript plugins are enabled` section.

**Comment by peterjuras** [comment]
Will the babel 8 release also include default handling of `.ts` and `.tsx` extensions? (see #8652). It's tagged with the milestone but this issue doesn't mention it.

**Comment by jridgewell** [comment]
Can we also take the chance to rename now-standard features still listed a `proposal-*`?

- babel-plugin-proposal-async-generator-functions
- babel-plugin-proposal-dynamic-import
- babel-plugin-proposal-export-namespace-from
- babel-plugin-proposal-json-strings
- babel-plugin-proposal-nullish-coalescing-operator
- babel-plugin-proposal-object-rest-spread
- babel-plugin-proposal-optional-catch-binding
- babel-plugin-proposal-optional-chaining
- babel-plugin-proposal-unicode-property-regex

**Comment by AviVahl** [comment]
> Currently they behave asynchronously if you pass them a callback, otherwise they behave synchronously. Only the callback-based version should be allowed.

How would such change be compatible with babel-register? Unless I'm missing something, `require.extensions` is synchronous.

**Comment by cpojer** [comment]
Hello! Thanks everyone for your hard work on Babel (8). I wanted to ask about your upgrade strategy. To this day, Babel 6 is still found in many dependency trees of projects that use Babel 7 just because some tools haven't been (properly) upgraded. With many tools pulling in Babel 7, what is your recommendation on adopting Babel 8 to avoid mixing Babel 7 and 8 (and 6)?

It seems that due to semver there will be an awkward period of time when some tooling will be updated to use Babel 8 and some won't, likely forcing projects to use both in their projects or hold off on upgrading for a long time altogether.

In the original post it was said that there will be a Babel 7 release with all the non-breaking changes, which is awesome! Is there any consideration to keep the version gap to a minimum? One approach could be further updates to Babel 7 that enable the breaking changes as opt-in configuration options and I'm sure there are others.

**Comment by nicolo-ribaudo** [comment]
It's not always possible, but in most cases we are making it so that something that works with Babel 8 can also work with Babel 7. Sometimes it works "out of the box" (for example for `Disallow using babel.transform, babel.transformFile, babel.transformFromAst, babel.parse, babel.loadOptions and babel.loadPartialConfig synchronously`), other times by enabling the corresponding option (for example, `Transforms JSX spread properties using object spread`).

**Comment by cpojer** [comment]
Keep in mind the impact of duplicating Babel 6 and Babel 7 was lower because the package names changed from the `babel-` prefix to using the `@babel` org. When Babel 8 comes out, it is likely that through hoisting rules every tool will pull in its own copy of Babel 7 or 8 within its sub-node_modules folder.

Babel 6 had a "bridge" mode that allowed installing version 7 of some packages like `babel-core` bridging to `@babel/core`.

**Comment by mohsen1** [comment]
Please add ts and tsx as default extensions. I spent a few hours to figure out babel-node need specifying extensions for it to work when importing ts files. @peterjuras

**Comment by kaicataldo** [comment]
I’m beginning to rethink our plan has been to release the `@babel/eslint-*` packages with Babel 8. I believe we’re ready to release those packages now - any further changes will become apparent with the myriad of integrations in the community. If we were to release these packages in the next v7 line release, we can start getting immediate feedback. One other possible benefit is that if we need to make any breaking changes we can get them out in the v8 release.

Thoughts?

**Comment by nicolo-ribaudo** [comment]
The next ESLint release has been in the prerelease state for too long. If it's already compatible with ESLint 7, I would be ok with releasing it in Babel 7.

**Comment by cpojer** [comment]
Question: could the official upgrade guide suggest how library authors can make their code compatible with both Babel 7 and 8, and then recommend them to use a semver range that works with either 7 or 8 until 7 is dropped at a later time?

I believe that could significantly reduce the gap in time when some tools are upgraded but others aren't.

**Comment by nicolo-ribaudo** [comment]
Yes!
1. We will check which Babel 7 plugins in this repository work with `@babel/core` 8, and relax their server ranges
2. The `@babel/core` API (used by Babel integrations such as Jest, Parcel, the webpack and rollup plugins, etc) will work both with Babel 7 and Babel 8. There is a breaking change (https://github.com/babel/babel/pull/11110), but the alternative already works with Babel 7.
3. If an open-source plugin author needs help to make their plugin work both with Babel 8 and Babel 7, I am happy to help on Slack (and probably also the rest of the team).

**Comment by cpojer** [comment]
Thanks for the reply @nicolo-ribaudo, that's really exciting as that will ease the transition 👍

**Comment by hzoo** [comment]
Was thinking through some ideas on how to make the migration easier (not for 1 project for the ecosystem) with @nicolo-ribaudo and @existentialism and I'll post an idea we had here: https://hackmd.io/2WTT3Z5HSZutVkk5biGKfA. (this is in addition to writing up a guide to migrate and codemods where possible)

Basically was thinking we could enable the breaking changes in a 7.x patch under an environmental flag, run like `BABEL_8_BREAKING babel` or with whatever integration so that any library/plugin can opt-into trying that locally and publish new versions ahead of us releasing anything. We would have to manage the if branches in the code and even tests + cleanup after but maybe that can help ease?

Other ideas welcome.

**Comment by cpojer** [comment]
@hzoo I think that's a great idea, thank you for thinking about this.

**Comment by Airkro** [comment]
How about drop support for `core-js 3` (in `preset-env`). We have `babel-plugin-polyfill-corejs3` now.

**Comment by nicolo-ribaudo** [comment]
That was one of my goals when I started working on the polyfill packages, but we should first do https://github.com/babel/rfcs/pull/2

**Comment by xiaoxiangmoe** [comment]
When will babel 8 release?

**Comment by TrySound** [comment]
Hi, I found cases like `<> ></>` are not errors in flow. Could you please clarify if it's also forbidden?
https://flow.org/try/#0JYWwDg9gTgLgBAKjgQwM5wEoFNkGN4BmUEIcA5FDvmQNwBQduEAdqvAMImTNbPwC8cABQBKOPwB8cAN504cSjACuUZnAA8UieoD0EugF8gA
This lint errors fails there
https://github.com/yannickcr/eslint-plugin-react/blob/master/docs/rules/no-unescaped-entities.md

**Comment by nicolo-ribaudo** [comment]
@TrySound It is: you can check `JSXTextCharacter` at https://facebook.github.io/jsx/.

**Comment by fabis94** [comment]
Would be nice to know when we can expect Babel 8. The OP says that it'll come out in 2020, but the year is soon coming to a close and there's no news yet.

**Comment by nicolo-ribaudo** [comment]
@fabis94 Sorry for the lack of communication about a precise schedule. My hope is to release Babel 8 alongside with 7.13 (so in about 2 months), since there isn't much left to do. However, please don't take this as an hard deadline since anything could happen :pray: 

@WillBeebe Starting from Babel 7.12.0, it will use `chokidar@3` if you are using a modern Node.js version :wink: https://github.com/babel/babel/pull/11560

**Comment by fabis94** [comment]
Any updates regarding this?

**Comment by nicolo-ribaudo** [comment]
Yes!
We wanted to work on this in a separate branch `next-8-dev`/`next-8-rebased`, but we found out that it was _really_ slowing us down. We are not starting to add these Babel 8 changes to `main`, so that it's easier for us to work on them and hopefully will help us finish this release in a reasonable time.

**Comment by nicolo-ribaudo** [comment]
No need to wait for Babel 8, we can add an alias now.

**Comment by rnarkk** [comment]
Would you mind updating the current progress of the task list? Some of them seem to have already been solved.

**Comment by Ontopic** [comment]
> Yes!
> We wanted to work on this in a separate branch `next-8-dev`/`next-8-rebased`, but we found out that it was _really_ slowing us down. We are now starting to add these Babel 8 changes to `main`, so that it's easier for us to work on them and hopefully will help us finish this release in a reasonable time.

So in order to play around with where things are at, will a @next or ^7.15 bring me the closest to where you guys are at?

**Comment by JLHwung** [comment]
> So in order to play around with where things are at, will a @next or ^7.15 bring me the closest?

7.15 has been published but it will not contain Babel 8 breaking changes, as required by semver. We don't have Babel 8 nightly build, but we have Babel 8 e2e test on popular libraries: https://github.com/babel/babel/blob/614b48678095746b83bbe517c4d6b30ba8cd5c04/.circleci/config.yml#L131 you can fork  Babel repo and run e2e test on your own library to see if Babel 8 will break. Feedbacks are always welcome.

**Comment by Ontopic** [comment]
Thanks! Long time since I last build Babel myself. Let's see the dev contributor experience nowadays :)

**Comment by Semigradsky** [comment]
> Bump Node Support to >=12.19

Should it be updated to `>=14`? [12's EOL is 2022-04-30](https://nodejs.org/en/about/releases/).

**Comment by nicolo-ribaudo** [comment]
Yes thanks, we are a few years late with this release! :upside_down_face:

**Comment by dlong500** [comment]
I hate to be that guy, and I know that ultimately a release is ready when it's ready, but given the timeframe here can we get an update on where things stand?  The Babel Roadmap page hasn't been updated since 2021 (at which point there was a note about how long it had already been taking and how the release was "closer than ever").

**Comment by Slapbox** [comment]
Friendly bump on the question above.

**Comment by csvan** [comment]
Hate to state the obvious, but the _best_ thing you can do to help issues like this is getting your employer - or yourself - to sponsor and otherwise contribute to open source. 

Many of the absolutely crucial open source projects - such as Babel - are developed by volunteers in their spare time, while large corporations and startups alike incorporate them into their commercial products without ever giving anything back. I am actively working in my organisation to get that changed, and I hope others will too.

**Comment by dlong500** [comment]
> Yes thanks, we are a few years late with this release! 🙃
> 
> (If it takes another year we can also drop Node.js 14 👀)

I guess Node 14 support can be dropped...

**Comment by ClearedFram3** [comment]
Now that 8 would be released in 2023, could dropping Flow, IE, and pre-ES2015 support be done?
- Flow is barely used
- IE is dead (thanks MS)
- ES2015 was 8 years ago!

**Comment by birkskyum** [comment]
Before this is out the gate, Node 16 will be well passed it's EOL as well https://nodejs.org/en/blog/announcements/nodejs16-eol

**Comment by nicolo-ribaudo** [comment]
Once https://github.com/babel/babel/pull/16416, https://github.com/babel/babel/issues/16443 and https://github.com/babel/babel/issues/16456 are done I plan to release a Babel 8.0.0-beta.0 version

**Comment by fisker** [comment]
`typescript-eslint` made some changes to their AST, [blog post](https://typescript-eslint.io/blog/announcing-typescript-eslint-v8-beta/), do we have a plan to align with it?

**Comment by MrSuddenJoy** [comment]
Whoever may be concerned about this: will `Babel 8` see the light of day anytime soon?

**Comment by SukkaW** [comment]
@nicolo-ribaudo 

Previously, the move of replacing `chalk` has been abandoned: https://github.com/babel/babel/pull/13783

Do you want to land this in Babel 8? I'd like to help with that.

**Comment by rschristian** [comment]
(Feel free to hide as this is fairly off topic)

@SukkaW In the small chance you hadn't yet seen, it looks like that actually landed in #16359. Came here looking myself after I saw usage of both `chalk` & `picocolors` in v7, which seemed a bit redundant. v8 should drop usage entirely _I believe_.

**Comment by michaelfaith** [comment]
Trying to track down a breaking change in the TypeScript types in `@babel/types`, and I traced it down to this commit, which was released as part of the v8 pre-release cycle.  Was this intended to go into 7.x?  Namely the change in shape of the `SourceLocation` and `Position` types?
https://github.com/babel/babel/commit/52244bf0834801684f1e55621b395f5af69f6a10#diff-8686732cac4b4c0405e0415ac800fb066039f1ae83786248e2e0c50924f17adf

**Comment by nicolo-ribaudo** [comment]
To clarify, the problem is that index should be optional for backwards compat? Happy to make that change if needed.

**Comment by michaelfaith** [comment]
> To clarify, the problem is that index should be optional for backwards compat? Happy to make that change if needed.

Yep, `index` in `Position` and `filename` in `SourceLocation`

**Comment by Dev-Re2906** [comment]
approve

**Comment by yoyo837** [comment]
Are there plans to convert the `@babel/eslint-parser/experimental-worker` to a non-experimental version in Babel 8?

**Comment by fisker** [comment]
Finally, we can close this?

**Comment by nicolo-ribaudo** [comment]
Yes! 🎉
